Docs: Added benchmarks for `torch.compile()` for vision models (#24748)
huggingface/transformers: docs/source/en/perf_torch_compile.md

> <!--Copyright 2022 The HuggingFace Team. All rights reserved.

Reviewer: ```suggestion
<!--Copyright 2023 The HuggingFace Team. All rights reserved.
```

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the

⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
rendered properly in your Markdown viewer.

-->

# Optimize inference using torch.compile()

> This guide aims to provide a benchmark on the inference speed-ups introduced with [`torch.compile()`](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html) for computer vision models in 🤗 Transformers.

Reviewer: Could we provide an HF Hub link of the vision models here and hyperlink "computer vision models in 🤗 Transformers" with that?

Author: You can really filter by modality but specific task, I can do image classification.

## Benefits of `torch.compile` 
   

> Depending on the model and the GPU, `torch.compile()` yields up to 30% speed-up during inference. To use `torch.compile()`, simply install any version of `torch` above 2.0. 

Reviewer: Requires CUDA 11.7 at least. I would also link this doc that has the detailed installation instructions of Torch 2.0: https://pytorch.org/get-started/pytorch-2.0/

Compiling a model takes time, so it's useful if you are compiling the model only once instead of every time you infer.
To compile any computer vision model of your choice, call `torch.compile()` on the model as shown below:

```diff
from transformers import AutoModelForImageClassification

model = AutoModelForImageClassification.from_pretrained(MODEL_ID).to("cuda")
+ model = torch.compile(model)
```

> `compile()` comes with multiple modes for compiling, which essentially differ in compilation time and inference overhead. `max-autotune` takes longer than `reduce-overhead` but results in faster inference. Default mode is fastest for compilation but is not as efficient compared to `reduce-overhead` for inference time. In this guide, we used the default mode. 

Reviewer: Where can the readers learn more about it? I'd provide a link to the official documentation.

> We benchmarked `torch.compile` with different computer vision models, tasks, types of hardware, and batch sizes on `torch` version 2.0.1.

Reviewer: Would it make sense to instead do it on the top trending or top downloaded models instead? Sorry, if it's already being done.

Author: it is, yes, e.g. https://huggingface.co/models?pipeline_tag=image-segmentation&sort=likes
https://huggingface.co/models?pipeline_tag=object-detection&sort=likes

I couldn't compile DETR for segmentation and discussed with Amy on it, it's due to some of the ops in the architecture. Some models are OOM after warmup for some GPUs and larger batch sizes, I only took OwlViT out of it.

## Benchmarking code 

Below you can find the benchmarking code for each task. We warm up the GPU before inference and take the mean time of 300 inferences, using the same image each time.

#### Image Classification with ViT

```python 
import torch
from PIL import Image
import requests
import numpy as np

> from transformers import AutoFeatureExtractor, AutoModelForImageClassification
> 
> url = 'http://images.cocodataset.org/val2017/000000039769.jpg'
> image = Image.open(requests.get(url, stream=True).raw)
> 
> extractor = AutoFeatureExtractor.from_pretrained("google/vit-base-patch16-224")
> model = AutoModelForImageClassification.from_pretrained("google/vit-base-patch16-224").to("cuda")
> 
> extracted_input = extractor(image, return_tensors='pt').to(device="cuda")

Reviewer 2: Feature extractors for the image processors are deprecated. It won't affect the results though - it'll have just loaded an image processor under the hood

```suggestion
from transformers import AutoImageProcessor, AutoModelForImageClassification

url = 'http://images.cocodataset.org/val2017/000000039769.jpg'
image = Image.open(requests.get(url, stream=True).raw)

processor = AutoImageProcessor.from_pretrained("google/vit-base-patch16-224")
model = AutoModelForImageClassification.from_pretrained("google/vit-base-patch16-224").to("cuda")

processed_input = processor(image, return_tensors='pt').to(device="cuda")
```

with torch.no_grad():
    _ = model(**extracted_input)

> 

Reviewer: The code doesn't have any snippets for compilation, though. 

I think it would make sense to provide complete scripts in the interest of transparency (maybe even as GitHub gists).

```

#### Object Detection with DETR

```python 

> from transformers import AutoFeatureExtractor, AutoModelForObjectDetection
> 
> extractor = AutoFeatureExtractor.from_pretrained("facebook/detr-resnet-50")
> model = AutoModelForObjectDetection.from_pretrained("facebook/detr-resnet-50").to("cuda")
> model = torch.compile(model)
> 
> texts = ["a photo of a cat", "a photo of a dog"]
> inputs = extractor(text=texts, images=image, return_tensors="pt").to("cuda")

Reviewer 2: ```suggestion
from transformers import AutoImageProcessor, AutoModelForObjectDetection

processor = AutoImageProcessor.from_pretrained("facebook/detr-resnet-50")
model = AutoModelForObjectDetection.from_pretrained("facebook/detr-resnet-50").to("cuda")
model = torch.compile(model)

texts = ["a photo of a cat", "a photo of a dog"]
inputs = processor(text=texts, images=image, return_tensors="pt").to("cuda")
```

processor = SegformerImageProcessor.from_pretrained("nvidia/segformer-b0-finetuned-ade-512-512")
model = SegformerForSemanticSegmentation.from_pretrained("nvidia/segformer-b0-finetuned-ade-512-512").to("cuda")
model = torch.compile(model)
inputs = processor(images=image, return_tensors="pt").to("cuda")

with torch.no_grad():
    _ = model(**seg_inputs)
```

Below you can find the list of the models we benchmarked.

- Image Classification: google/vit-base-patch16-224, microsoft/beit-base-patch16-224-pt22k-ft22k
- Image Segmentation: nvidia/segformer-b0-finetuned-ade-512-512
- Object Detection: google/owlvit-base-patch32, facebook/detr-resnet-50

Below you can find inference durations in milliseconds for each model with and without `compile()`. Note that OwlViT results in OOM in larger batch sizes.

### A100 (batch size: 1)

| **Task/Model** | **torch 2.0 - <br>no compile** | **torch 2.0 - <br>compile** |
|:---:|:---:|:---:|
| Image Classification/ViT | 9.325 | 7.584 | 
| Image Segmentation/Segformer | 11.759 | 10.500 |
| Object Detection/OwlViT | 24.978 | 18.420 |
| Image Classification/BeiT | 11.282 | 8.448 | 

> | Object Detection/DETR | 34.619 | 19.040 |

Reviewer: Hmm seems like the compilation gains are not very prudent for smaller models.

### A100 (batch size: 4)

| **Task/Model** | **torch 2.0 - <br>no compile** | **torch 2.0 - <br>compile** |
|:---:|:---:|:---:|
| Image Classification/ViT | 14.832 | 14.499 | 
| Image Segmentation/Segformer | 18.838 | 16.476 |
| Image Classification/BeiT | 13.205 | 13.048 | 
| Object Detection/DETR | 48.657 | 32.418|

### A100 (batch size: 16)

| **Task/Model** | **torch 2.0 - <br>no compile** | **torch 2.0 - <br>compile** |
|:---:|:---:|:---:|
| Image Classification/ViT | 40.944 | 40.010 | 
| Image Segmentation/Segformer | 37.005 | 31.144 |
| Image Classification/BeiT | 41.854 | 41.048 | 

> | Object Detection/DETR | 164.382 | 161.902 |

Reviewer: No speedup at all except for SegFormer? 

Have you tried running a Benchmark for CNN models from Transformers (such as ConvNeXt)? Would be cool to understand if non-CNN models are not better candidates for `torch.compile()`.

Author: yep, it's a bit weird for larger batch sizes :/

Depending on the model and the GPU, `torch.compile()` yields up to 30% speed-up during inference. To use `torch.compile()`, simply install any version of `torch` above 2.0. 

Note that compiling the model takes time, so it's useful so long as you are compiling the model once and inferring repeatedly instead of compiling for each time you infer.
To compile any computer vision model of your choice, simply call `torch.compile()` on the model like below.

```diff
from transformers import AutoModelForImageClassification

model = AutoModelForImageClassification.from_pretrained(MODEL_ID).to("cuda")
+ model = torch.compile(model)
```

`compile()` comes with multiple modes for compiling, which essentially differ in compilation time and inference overhead. `max-autotune` takes longer than `reduce-overhead` but results in faster inference. Default mode is fastest for compilation but is not as efficient compared to `reduce-overhead` for inference time. In this guide, we have used default mode. 

We have benchmarked `torch.compile` with different computer vision models, tasks, different types of hardware and batch sizes. `torch` version used in this benchmark is 2.0.1.

### Benchmarking code 

> Below you can find the benchmarking code for each task. We warm the GPU up before the inference and take the mean of 300 inferences. 
> We used the same image for all inferences.

Reviewer 3: ```suggestion
Below you can find the benchmarking code for each task. We warm up the GPU before inference and take the mean time of 300 inferences, using the same image each time.
```

Below you can find inference durations in milliseconds for each model with and without `compile()`. Note that OwlViT results in OOM in larger batch sizes.

### A100 (batch size: 1)

| **Task/Model** | **torch 2.0 - <br>no compile** | **torch 2.0 - <br>compile** |
|:---:|:---:|:---:|
| Image Classification/ViT | 9.325 | 7.584 | 
| Image Segmentation/Segformer | 11.759 | 10.500 |
| Object Detection/OwlViT | 24.978 | 18.420 |
| Image Classification/BeiT | 11.282 | 8.448 | 
| Object Detection/DETR | 34.619 | 19.040 |

### A100 (batch size: 4)

| **Task/Model** | **torch 2.0 - <br>no compile** | **torch 2.0 - <br>compile** |
|:---:|:---:|:---:|
| Image Classification/ViT | 14.832 | 14.499 | 
| Image Segmentation/Segformer | 18.838 | 16.476 |
| Image Classification/BeiT | 13.205 | 13.048 | 
| Object Detection/DETR | 48.657 | 32.418|

### A100 (batch size: 16)

| **Task/Model** | **torch 2.0 - <br>no compile** | **torch 2.0 - <br>compile** |
|:---:|:---:|:---:|
| Image Classification/ViT | 40.944 | 40.010 | 
| Image Segmentation/Segformer | 37.005 | 31.144 |
| Image Classification/BeiT | 41.854 | 41.048 | 
| Object Detection/DETR | 164.382 | 161.902 |

> ### V100 (batch size: 1) (DONE)

Reviewer 3: ```suggestion
### V100 (batch size: 1)
```

<!--Copyright 2023 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the

⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
rendered properly in your Markdown viewer.

-->

# Optimize inference using torch.compile()

This guide aims to provide a benchmark on the inference speed-ups introduced with [`torch.compile()`](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html) for [computer vision models in 🤗 Transformers](https://huggingface.co/models?pipeline_tag=image-classification&library=transformers&sort=trending).

## Benefits of `torch.compile` 
   

> Depending on the model and the GPU, `torch.compile()` yields up to 30% speed-up during inference. To use `torch.compile()`, simply install any version of `torch` above 2.0. 

Reviewer 3: Not sure why `torch.compile` is properly rendered in code-format, can you maybe try to take the backticks out of the section header to see if that helps?

Author: sure!

Compiling a model takes time, so it's useful if you are compiling the model only once instead of every time you infer.
To compile any computer vision model of your choice, call `torch.compile()` on the model as shown below:

```diff
from transformers import AutoModelForImageClassification

model = AutoModelForImageClassification.from_pretrained(MODEL_ID).to("cuda")
+ model = torch.compile(model)
```

`compile()` comes with multiple modes for compiling, which essentially differ in compilation time and inference overhead. `max-autotune` takes longer than `reduce-overhead` but results in faster inference. Default mode is fastest for compilation but is not as efficient compared to `reduce-overhead` for inference time. In this guide, we used the default mode. You can learn more about it [here](https://pytorch.org/get-started/pytorch-2.0/#user-experience).

We benchmarked `torch.compile` with different computer vision models, tasks, types of hardware, and batch sizes on `torch` version 2.0.1.

## Benchmarking code 

Below you can find the benchmarking code for each task. We warm up the GPU before inference and take the mean time of 300 inferences, using the same image each time.

> #### Image Classification with ViT

Reviewer 3: These should be level 3 now

```suggestion
### Image Classification with ViT
```

```python 
import torch
from PIL import Image
import requests
import numpy as np
from transformers import AutoImageProcessor, AutoModelForImageClassification

url = 'http://images.cocodataset.org/val2017/000000039769.jpg'
image = Image.open(requests.get(url, stream=True).raw)

processor = AutoImageProcessor.from_pretrained("google/vit-base-patch16-224")
model = AutoModelForImageClassification.from_pretrained("google/vit-base-patch16-224").to("cuda")
model = torch.compile(model)

processed_input = processor(image, return_tensors='pt').to(device="cuda")

with torch.no_grad():

>     _ = model(**extracted_input)

Reviewer 2: ```suggestion
    _ = model(**processed_input)
```

```

#### Object Detection with DETR

```python 
from transformers import AutoImageProcessor, AutoModelForObjectDetection

processor = AutoImageProcessor.from_pretrained("facebook/detr-resnet-50")
model = AutoModelForObjectDetection.from_pretrained("facebook/detr-resnet-50").to("cuda")
model = torch.compile(model)

texts = ["a photo of a cat", "a photo of a dog"]
inputs = processor(text=texts, images=image, return_tensors="pt").to("cuda")

with torch.no_grad():
    _ = model(**inputs)
```

#### Image Segmentation with Segformer

```python 
from transformers import SegformerImageProcessor, SegformerForSemanticSegmentation

processor = SegformerImageProcessor.from_pretrained("nvidia/segformer-b0-finetuned-ade-512-512")
model = SegformerForSemanticSegmentation.from_pretrained("nvidia/segformer-b0-finetuned-ade-512-512").to("cuda")
model = torch.compile(model)

> inputs = processor(images=image, return_tensors="pt").to("cuda")

Reviewer 2: ```suggestion
seg_inputs = processor(images=image, return_tensors="pt").to("cuda")
```

with torch.no_grad():
    _ = model(**seg_inputs)
```

Below you can find the list of the models we benchmarked.

> **Image Classification** 

Reviewer: Can we hyperlink these models or will that be an overkill?

Author: I just did

- google/vit-base-patch16-224 
- microsoft/beit-base-patch16-224-pt22k-ft22k 
- facebook/convnext-large-224
- microsoft/resnet-50

**Image Segmentation** 
- nvidia/segformer-b0-finetuned-ade-512-512
- facebook/mask2former-swin-tiny-coco-panoptic
- facebook/maskformer-swin-base-ade
- google/deeplabv3_mobilenet_v2_1.0_513

**Object Detection** 
- google/owlvit-base-patch32
- facebook/detr-resnet-101
- microsoft/conditional-detr-resnet-50

Below you can find visualization of inference durations with and without `torch.compile()` and percentage improvements for each model in different hardware and batch sizes. 

> ![Duration Comparison on A100 with Batch Size of 1](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/torch_compile/A100_1_duration.png)

Reviewer 2: It would be cool to have this graph with the % improvements for each model across batch sizes i.e. different coloured bars for each batch size

Reviewer 3: Maybe it'd also be nicer to display the graphs side by side like [this](https://play.tailwindcss.com/5GFFWYX51m)?

Author: I added them

![Percentage Improvement on A100 with Batch Size of 1](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/torch_compile/A100_1_percentage.png)

![Duration Comparison on V100 with Batch Size of 1](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/torch_compile/v100_1_duration.png)

![Percentage Improvement on T4 with Batch Size of 4](https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/torch_compile/T4_4_percentage.png)

Below you can find inference durations in milliseconds for each model with and without `compile()`. Note that OwlViT results in OOM in larger batch sizes.

### A100 (batch size: 1)

> | **Task/Model** | **torch 2.0 - <br>no compile** | **torch 2.0 - <br>compile** |

Reviewer 2: It it's easy to do, it might be nice to add the % differences to the table too. I don't feel strongly about this though, so happy for you to reject as an idea :)

Author: sure :)

| Image Classification/BeiT | 52.007 | 51.354 | 
| Object Detection/DETR | 277.850 | 268.003 |
| Image Classification/ConvNeXT | 119.259 | 105.580 | 
| Image Classification/ResNet | 13.039 | 11.388 |
| Image Segmentation/Mask2former | 201.540 | 184.670 |
| Image Segmentation/Maskformer | 764.052 | 711.280 | 
| Image Segmentation/MobileNet | 74.289 | 48.677 |
| Object Detection/Resnet-101 | 421.859 | 357.614 |
| Object Detection/Conditional-DETR | 289.002 | 226.945 |

### T4 (batch size: 16)

| **Task/Model** | **torch 2.0 - <br>no compile** | **torch 2.0 - <br>compile** |
|:---:|:---:|:---:|
| Image Classification/ViT | 163.914 | 160.907 | 
| Image Segmentation/Segformer | 192.412 | 163.620 |
| Image Classification/BeiT | 188.978 | 187.976 | 
| Object Detection/DETR | OOM | OOM |
| Image Classification/ConvNeXT | 422.886 | 388.078 | 
| Image Classification/ResNet | 44.114 | 37.604 |
| Image Segmentation/Mask2former | 756.337 | 695.291 |
| Image Segmentation/Maskformer | 2842.940 | 2656.88 | 
| Image Segmentation/MobileNet | 299.003 | 201.942 |
| Object Detection/Resnet-101 |  1619.505 | 1262.758 | 
| Object Detection/Conditional-DETR | 1137.513 | 897.390|

## PyTorch Nightly

> We also benchmarked on PyTorch nightly (2.1.0dev) and observed improvement in latency both for uncompiled and compiled models.

Reviewer 2: @merveenoyan Awesome - thanks for running and adding info on nightly! For the tables below, I think there's a column missing, as there's 5 headers and 4 cols of info.

> We also benchmarked on PyTorch nightly (2.1.0dev) and observed improvement in latency both for uncompiled and compiled models.

Reviewer 2: Small nit - could you mention which commit you were running from for the nightly tests?

Author: I directly downloaded the wheel so I can give a link to that if it's ok

Reviewer 2: Yep - I think so. As long as it's possible to know which nightly is being referred to.

Author: Added link to wheel.
